bokeh/colors/util.py

```diff
@@ -119,12 +119,40 @@ def __init__(self, name: str, r: int, g: int, b: int) -> None:
         self.name = name
         super().__init__(r, g, b)
 
+    @classmethod
+    def find(cls, name: str) -> NamedColor:
+        # If name not recognised, raises ValueError
+        index = cls.__all__.index(name)
+        return cls.colors[index]
+
     def to_css(self) -> str:
         '''
 
         '''
         return self.name
 
+def _color_as_rgb_hex(color: str) -> str:
+    ''' Convert string color to RGB hex string.
+
+    Color may be a named color such as "blue", or a hex RGB(A) string such as
+    "#FFAA00", "#FFAA00FF", "#FA0" or "#FA0F". Returned string is always of
+    the form "#RRGGBB" with 2 characters for each of the R, G and B components
+    and no alpha component.
+
+    '''
+
+    if color.startswith("#"):
+        if len(color) in (7, 9):
+            return color[:7].upper()
+        elif len(color) in (4, 5):
+            return f"#{color[1]*2}{color[2]*2}{color[3]*2}".upper()
+        else:
+            raise ValueError(f"Unable to parse hex color string: '{color}'")
```

Comment:
Given the logic of this function and its usage, I would let `Color` or `RGB` classes to do the parsing and let the rest of the code work on `RGB` instances, instead of strings.

Reply:
I've followed this recommendation so now there is an `RGB.from_hex_string()` function which parses the string, and a `NamedColor.from_string()` which tries both `RGB.from_hex_string()` and `NamedColor.find()` in turn so it accepts color names as well as hex RGBA strings. This feels like better design as all access to the `NamedColor` class variables is within the class.

`varying_alpha_palette()` is now a pretty simple function. For the output RGBA hex strings I am calling `RGB.to_hex()` which ignores the `alpha` component and appending the interpolated `alpha` to it. It might be better to modify `RGB.to_hex()` to append `alpha` as well, but it would need to have a new boolean kwarg such as `want_alpha` or equivalent.